src/mol-math/geometry/lookup3d/grid.ts

```diff
@@ -40,6 +40,40 @@ class GridLookup3DImpl<T extends number = number> implements GridLookup3D<T> {
         return ret;
     }
 
+    nearest(x: number, y: number, z: number): { index: number, squaredDistance: number } | undefined {
+        if (!OrderedSet.size(this.ctx.grid.data.indices)) return undefined;
+        this.ctx.x = x;
+        this.ctx.y = y;
+        this.ctx.z = z;
+        const radiusIncrement = 0.1; // how to choose a good increment?
+        const startingRadius = this.distanceTo(x, y, z);
+        this.ctx.radius = startingRadius < 0 ? radiusIncrement : startingRadius + radiusIncrement;
+        this.ctx.isCheck = false;
+        const result = this.result;
+        query(this.ctx, result);
+        while (!result.count) { // necessary to check if grid is empty to avoid infinite loop
+            this.ctx.radius += radiusIncrement;
+            query(this.ctx, result);
+        }
+        const { indices, squaredDistances } = result;
+        let index = indices[0], nearestDist = squaredDistances[0];
```

Comment:
Instead of this, how about you find the voxel the point is in and then recursively visit neighboring cells to find the closest point? You always have to visit at least 1 layer of closest voxels. If you don't find the closest point there, you increase the threshold. There are some clever ways to do this too (like you can keep queue of voxels you need to visit represented as integer indices to avoid array allocations).

Reply:
> Instead of this, how about you find the voxel the point is in and then recursively visit neighboring cells to find the closest point? You always have to visit at least 1 layer of closest voxels. If you don't find the closest point there, you increase the threshold. 

Yes, I could do something like this. But in case the point is outside the grid this could lead to problems in my understanding. Expanding from the nearest voxel/s instead of the original center could exclude nearer points from the result. Should we use two different methods in case the point is inside or outside the grid?

Reply:
Expanding from the nearest voxel should work as well as you can expand in a way similar to how you expand your radius here. You just keep the distance to the furthest current included point and based on the voxel position (e.g. center of the voxel, you can stop the search).

Reply:
@dsehnal i tried to draw what I mean. This is a 2D example but I guess it should be the same in 3D. Also please let me know if there is something conceptually wrong (for example I am unsure about the boundary sphere radius)

There are two expansions in gray, one from the external point origin and the other from the nearest voxel. In this configuration the nearest point respect to the nearest voxel isn't the same of the external point expansion.

![image](https://user-images.githubusercontent.com/20627977/188153207-8f6417fd-9c02-40d2-a54e-80de5b991e28.png)

Reply:
The method I am suggesting will still work. You do not stop searching the 1st time you find a point, you stop searching when all voxels that have not been visited are further away then the closest point you've found:

![image](https://user-images.githubusercontent.com/21145776/188182533-e81f4530-9f2f-4e8e-97b6-2864520f9391.png)

You can do this using breath-first search (with termination condition also based on the on the currently found closest point) of the voxels starting from the closest one.

Reply:
To clarify a bit more, you check against the distance to the query point, not distance to the closest voxel.

Reply:
Ok thanks, I think I understood now. 

Regarding the way to find the nearest voxel do you have any ideas? Should I evaluate the position of each voxel and keep the nearest one? Or maybe something else?

Instead regarding the layer expansion I was thinking to keep track of already visited voxels ( in a map maybe? ) and move the current layer voxels in each direction avoiding already visited voxels. Is this a good approach?

Reply:
I think just doing ``min(max(0, floor(x / cell_size_x)), n_x)`` for each dimension for the initial cell should be a reasonable estimate.

For points outside the bounding sphere, doing intersection between the ray going from the point to the center of the bounding sphere and the sphere and "snap" that point to the closest cell from the step above might provide a better estimate.

Even a better estimate for outside points could be doing intersection between the faces of the grid box and the ray from B. Although this might be somewhat harder to compute.

Reply:
@dsehnal with the last commit I implemented a method nearestPointToVec in box3D. It should return the very nearest point in the box respect to a random vector. This is the concept:
- The method first searches for the point to be inside the box
- If not, it makes projections on all planes of the box. If projection is inside the box (on the surface of the box) the distance and the vector are kept to make comparisons.
- If at least one projection is on the surface of the box it returns the projection with minimum distance from the point.
- If no projections on planes are within the box boundaries, it makes projections on all 12 segments of the box and evaluate distance only if projection is inside the segment like before.
- If everything fails it looks at all vertices of the box and returns the vertex with minimum distance.

Reply:
I think this is too complicated. All you need is an intersection between a ray from the query point and the center of the bounding sphere/box. Here is a simple method https://gamedev.stackexchange.com/a/18459. If this fails, you can simply use the closest vertex of the box as a starting point.

Reply:
>Even a better estimate for outside points could be doing intersection between the faces of the grid box and the ray from B. Although this might be somewhat harder to compute.

I was just trying to apply this method you mentioned. But searching online I realized it was necessary to search also for projections on edges and on vertices.


>Here is a simple method https://gamedev.stackexchange.com/a/18459. If this fails, you can simply use the closest vertex of the box as a starting point.

Sorry, this is not my field.. a few questions regarding this method.
- It isn't clear to me how to get the intersection point from this?
- I'm not sure how could it fail since the ray is directed towards the bounding box center?

Reply:
Also an other question, 
Does molstar always uses AABB boxes type? Even for instances with transformed coordinates for example?

Reply:
> I'm not sure how could it fail since the ray is directed towards the bounding box center?

Yeah, this can't happen (but I am always paranoid about these things).

> Does molstar always uses AABB boxes type? Even for instances with transformed coordinates for example?

You query the transformed units by applying an inverse transform on the query point, so it's always AABB.